vembrane/modules/filter.py

```diff
@@ -62,46 +62,89 @@ def add_subcommmand(subparsers):
+        try:
+            annotations = record.info[ann_key]
+        except KeyError:
+            annotations = [""]
+        #  … and only keep the annotations where the expression evaluates to true
+        filtered_annotations = [
+            annotation for annotation in annotations if env.evaluate(annotation)
+        ]  # TODO: if keep_unmatched, that could be "any(iterator)" instead of list
+
+        if not keep_unmatched and (len(annotations) != len(filtered_annotations)):
+            # update annotations if they have actually been filtered
+            record.info[ann_key] = filtered_annotations
+
+        return record, len(filtered_annotations) > 0
+    else:
+        # otherwise, the annotations are irrelevant w.r.t. the expression,
+        # so we can omit them
+        return record, env.evaluate()
 
 
 def filter_vcf(
     vcf: VariantFile,
     expression: str,
     ann_key: str,
     keep_unmatched: bool = False,
+    preserve_order: bool = False,
 ) -> Iterator[VariantRecord]:
 
     env = Environment(expression, ann_key, vcf.header)
 
+    events = set()
+    info_keys = set(vcf.header.info.keys())
+
     record: VariantRecord
     for idx, record in enumerate(vcf):
-        env.update_from_record(idx, record)
-        if env.expression_annotations():
-            # if the expression contains a reference to the ANN field
-            # get all annotations from the record.info field
-            # (or supply an empty ANN value if the record has no ANN field)
-            try:
-                annotations = record.info[ann_key]
-            except KeyError:
-                annotations = [""]
-            #  … and only keep the annotations where the expression evaluates to true
-            filtered_annotations = [
-                annotation for annotation in annotations if env.evaluate(annotation)
-            ]
-            if not filtered_annotations:
-                # skip this record if filter removed all annotations
-                continue
-            elif not keep_unmatched and (len(annotations) != len(filtered_annotations)):
-                # update annotations if they have actually been filtered
-                record.info[ann_key] = filtered_annotations
-            yield record
-        else:
-            # otherwise, the annotations are irrelevant w.r.t. the expression,
-            # so we can omit them
-            if env.evaluate():
+        record, passed = test_and_update_record(
+            env, idx, record, ann_key, keep_unmatched
+        )
+        if passed:
+            is_bnd = "SVTYPE" in info_keys and record.info.get("SVTYPE", None) == "BND"
+            if is_bnd:
+                event = record.info.get("EVENT", None)
+                events.add(event)
+            elif not preserve_order:
+                # if preserver order, we will output everything in the second pass *
                 yield record
+
+    if len(events) > 0:
+        # perform a second pass
```

Comment:
```suggestion
        # perform a second pass if the first pass filtered breakend (BND) events
        # since these are compound events which have to be filtered jointly
```